### PR TITLE
[#139] fix event concurrency issues

### LINKED ISSUE #139: Port iceoryx1 waitset bitset to iceoryx2 to handle overflowing events
## Required information

Currently, the event cannot handle trigger overflows and the notifier receives connection failures when the listener goes out of scope while the notifier is triggering. Furthermore, the tests in `service_event_tests.rs` named `concurrent***` segfault sometimes.

To address the issues:
* [ ] All concept shall be based on the bitset from iceoryx1
* [ ] All ignored tests from `service_event_tests.rs` must be unignored 
* [ ] Introduce event tests with many reconnecting listeners for many reconnecting notifiers.

### PR DESCRIPTION
**Pull request by elfenpiff**
## Pre-Review Checklist for the PR Author

1. [x] Add sensible notes for the reviewer
1. [x] PR title is short, expressive and meaningful
1. [x] Relevant issues are linked in the [References](#references) section
1. [x] Every source code file has a copyright header with `SPDX-License-Identifier: Apache-2.0 OR MIT`
1. [x] Branch follows the naming format (`iox2-123-introduce-posix-ipc-example`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`[#123] Add posix ipc example`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [x] Assign PR to reviewer
1. [x] All checks have passed (except `task-list-completed`)

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/concepts/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx2/blob/main/doc/release-notes/iceoryx2-unreleased.md

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Unit tests have been written for new behavior
- [x] Public API is documented
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

Closes #139

### CODE REVIEW

File: iceoryx2/src/service/zero_copy.rs [hunk 1]
```diff
@@ -52,9 +52,9 @@ impl crate::service::Service for Service {
     type ServiceNameHasher = hash::sha1::Sha1;
     type SharedMemory = shared_memory::posix::Memory<PoolAllocator>;
     type Connection = zero_copy_connection::posix_shared_memory::Connection;
-    #[cfg(not(any(target_os = "macos", target_os = "windows")))]
-    type Event = event::sem_bitset_posix_shared_memory::Event;
-    #[cfg(any(target_os = "macos", target_os = "windows"))]
+    //#[cfg(not(any(target_os = "macos", target_os = "windows")))]
+    //type Event = event::sem_bitset_posix_shared_memory::Event;
+    //#[cfg(any(target_os = "macos", target_os = "windows"))]
```
Comment (elBoberido) [inline iceoryx2/src/service/zero_copy.rs:57]: Can this be removed?

File: iceoryx2/tests/service_event_tests.rs [hunk 2]
```diff
@@ -490,30 +491,30 @@ mod service_event {
...
         let _watch_dog = Watchdog::new();
 
-        const NUMBER_OF_LISTENER_THREADS: usize = 2;
-        const NUMBER_OF_NOTIFIER_THREADS: usize = 2;
-        const NUMBER_OF_ITERATIONS: usize = 50;
-        const EVENT_ID: EventId = EventId::new(558);
+        let number_of_listener_threads =
+            (SystemInfo::NumberOfCpuCores.value() / 2).clamp(2, usize::MAX);
+        let number_of_notifier_threads =
+            (SystemInfo::NumberOfCpuCores.value() / 2).clamp(2, usize::MAX);
```
Comment (elBoberido) [inline iceoryx2/tests/service_event_tests.rs:500]: I would double the number of thread instead of cutting them in halve ... or maybe do tests with both.

**Review by elBoberido — APPROVED** [review]
Looks good. The commented out code could be removed, either in this PR or a follow up PR.

File: iceoryx2/src/service/zero_copy.rs (hunk 1 above)
Comment (elfenpiff) [reply iceoryx2/src/service/zero_copy.rs:57]: Whoopsie, oversight.

**Review by elBoberido — DISMISSED** [review]

**Review by elBoberido — APPROVED** [review]

**Review by elBoberido — APPROVED** [review]

**Pull request merged**